pkg/autodiscovery/listeners/types.go

```diff
@@ -14,14 +14,21 @@ import (
 // ID is the representation of the unique ID of a Service
 type ID string
 
+// ContainerPort represents a network port in a Service. Note that not all Services support
+// named network ports.
+type ContainerPort struct {
+	ContainerPort int
+	Name          string
+}
+
 // Service represents an application we can run a check against.
 // It should be matched with a check template by the ConfigResolver using the
 // ADIdentifiers field.
 type Service interface {
 	GetID() ID                            // unique ID
 	GetADIdentifiers() ([]string, error)  // identifiers on which templates will be matched
 	GetHosts() (map[string]string, error) // network --> IP address
-	GetPorts() ([]int, error)             // network ports
+	GetPorts() ([]ContainerPort, error)   // network ports
```

Comment:
An alternative approach is to let  the `Service` interface `ResolvePort([]byte) []byte` so that the implementation details about how a port is resolved (either by name or index for k8s) remains hidden. This would duplicate a bit of code but it might make sense if resolving a port by name **only** makes sense for k8s.

Reply:
Resolving ports by name should not be specific to Kubernetes. Although, I'm not sure why this implementation would be specific to Kubernetes - The notion of `Service` here is different from the kubernetes service.

Reply:
A `Service` in this context is just an interface but I was wondering if the logic for resolving port names makes more sense as an implementation detail since only k8s flavoured implementations of a `Service` have the concept of canonical port names (that is `PodContainerService` and `DockerKubeletService`). The `DockerService` and `ECSService` won't use the `Name` field.

This is a minor detail overall and I am comfortable leaving this as is.